Added inital complex-chart example

amilajack opened the pull request:
This PR was made as a follow up to https://github.com/uber/react-vis/pull/233, which was too large of a PR to be reviewed. This is an example of what other following examples should look like. Comments on the example in this PR will be reflected in examples that will be written later.

Here's a summary of what this PR adds for the complex-chart example:
- [x] Creates a `./examples` directory for where the current and future examples will be located
- [x] `babel-loader` integration. https://github.com/uber/react-vis/pull/233 used `buble-loader`. I was having module loader issues with this today so I used `babel-loader` as a (temporary?) replacement.
- [x] Aliased `react-vis` to `'../../dist'`. `babel-plugin-module-resolver` was preventing webpack from resolving the `index` alias created by `babel-plugin-module-resolver`
- [x] Created a `./examples/examples.scss`, borrowed from the original `./src/example/main.scss`
- [x] Borrowed `./examples/complex-chart/complex-chart-example.js` from `./showcase/plot/complex-chart.js`
- [x] Add `react-addons-shallow-compare` to `./examples/complex-chart/package.json`'s `dependencies`
- [x] Add `react` to `./examples/complex-chart/package.json`'s `dependencies`

mcnuttandrew commented on examples/complex-chart/README.md:
```diff
@@ -0,0 +1,10 @@
+react-vis legends
```
why is this section titled legends?

amilajack replied:
This was a mistake. I'll fix and rebase :)

mcnuttandrew commented on examples/complex-chart/app.js:
```diff
@@ -0,0 +1,66 @@
+//
+// When using >= 0.15.0, render with ReactDOM.render(). Otherwise, use React.render()
+
+const el = document.createElement('div');
```
maybe a little cleaner to add the div ahead of time to the index.html page and access it via a className? not wedded to that solution though

amilajack replied:
>  I tried patching into this PR and followed the instructions and got an error:

It requires that the `./dist` directory be built. When doing a fresh clone, this doesnt happen. Ideally, we could use webpack to alias `react-vis` => `./src` instead of `./dist` but I was having issues with `babel-plugin-module-resolver`.

Apercu commented on examples/complex-chart/app.js:
```diff
@@ -0,0 +1,66 @@
+
+const el = document.createElement('div');
+
+const render = isReactDOMSupported() ?
```
We don't need to support react 0.14, please use the `ReactDOM` render

Apercu commented on examples/complex-chart/app.js:
```diff
@@ -0,0 +1,66 @@
+export default class App extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {width: window.innerWidth};
```
Prefer the class property for of state, and add your listener in `componentDidMount`.

amilajack replied:
Should we use `componentWillMount` instead to prevent possible layout trashing?

Apercu commented on examples/complex-chart/app.js:
```diff
@@ -0,0 +1,66 @@
+
+import ComplexChartExample from './complex-chart-example';
+import '../examples.scss';
+import '../../src/main.scss';
```
You could also make `examples.scss` includes the main one, like in the showcase.

Apercu commented on examples/complex-chart/app.js:
```diff
@@ -0,0 +1,66 @@
+import React, {Component} from 'react';
+import window from 'global/window';
+import document from 'global/document';
+import {isReactDOMSupported} from '../../src/utils/react-utils';
```
The module resolver was specifically added to avoid such things. Also, you shouldn't require source files, but rather use `react-vis`.

Apercu commented on examples/complex-chart/complex-chart-example.js:
```diff
@@ -0,0 +1,199 @@
+  constructor(props) {
+    super(props);
+    const totalValues = Math.random() * 50;
+    this.state = {
```
ditto, class property

Apercu replied:
Class property instead?

Apercu commented on examples/complex-chart/complex-chart-example.js:
```diff
@@ -0,0 +1,199 @@
+        }
+      ]
+    };
+    this._nearestXHandler = this._nearestXHandler.bind(this);
```
If you use class properties for your methods, you won't have to bind all of them in the constructor, much cleaner.

amilajack replied:
I think we should use autobound methods. Is that what you're referring to?

Apercu replied:
If that's how you call class properties ;)

Apercu commented on examples/complex-chart/complex-chart-example.js:
```diff
@@ -0,0 +1,199 @@
+
+        <div className="chart">
+          <FlexibleXYPlot
+            animation={true}
```
Don't need `true`, just pass the animation prop

amilajack replied:
I see. Just to confirm, we want something like this:
```js
<FlexibleXYPlot
      animation
      {/* etc */} />
```
?

Apercu replied:
Yes

Apercu commented on examples/complex-chart/webpack.config.js:
```diff
@@ -0,0 +1,48 @@
+  //       section. This is example specific
+  resolve: {
+    alias: {
+      'react-vis': resolve('../../dist'),
```
This should not be done, remove all aliases to parent directories.

amilajack replied:
By removing the `react-vis` alias to `../../dist` dir, the examples will be built against the latest reactt-vis npm package instead of the source. Is that okay?

Apercu replied:
Yes, we want something clonable and runnable on its own. The dist would not be there anyway until you install all the dependencies and run the build.

Apercu commented on examples/complex-chart/webpack.config.js:
```diff
@@ -0,0 +1,48 @@
+  module: {
+    rules: [
+      {
+        test: /\.(sass|scss)$/,
```
We're only using `.scss` files

Apercu commented on examples/examples.scss:
```diff
@@ -0,0 +1,287 @@
+@import url(https://fonts.googleapis.com/css?family=Sintony:400,700);
```
This import should be extracted and put as a stylesheet in `index.html`

amilajack commented:
Should the examples support configurations so that they're easily migratable or that they should work only for these examples? For example, in the `webpack.config.js`, I made sure that the configuration supports both `sass` and less files. Another example is supporting react 0.14.0 and 0.15.0. A lot of the comments that @Apercu made were related to making catering the configuration only towards the examples, which has its benefits. I tried to make the examples work out of the box to help save consumers of the examples from some js-fatigue.

Which direction should we take?

Apercu commented:
I don't think providing loaders that are not even used because we think some users (less? maybe 15%?) might need them is a good idea. The example should be something clonable and runnable easily, without overthinking what the users might want. Most of them will have their own configuration anyway, and if they don't, they'll take what we give to tweak it after.

You should depend on react 15 and react-dom.

Also, I have a feeling creating multiples packages/webpack configs might not be a good idea since there would be a lot of code duplication. We have to think of something that would work nicely

amilajack commented:
> We have to think of something that would work nicely

Yup, definitly. I'm not sure how we can make it cloneable without including one config per example. @mcnuttandrew Thoughts on this?

Apercu commented:
@amilajack You might want to look at what I've [did](https://github.com/uber/deck.gl/blob/master/exhibits/webpack/package.json#L6) on deck.gl to prevent this issue.

amilajack commented:
Rebased the PR based on reviews:
- [x] Symlinked `webpack.config.js` and `index.html` to parent avoid duplication
- [x] Autobound methods to avoid `*.method.bind()` boilerplate
- [x] Added missing `dependencies` (react, react-dom)
- [x] Removed unnecessary `constructor()` in `examples/complex-chart/app.js`
- [x] Imported `main.scss` from `examples.scss`
- [x] Refactored font retrieval from `examples.scss` to avoid request waterfall
- [x] Dropped react 0.14.0 support
- [x] Other minor/miscellaneous changes

amilajack commented:
Question: Should the `./src/main.scss` be symlinked to `./examples/main.scss`?

Apercu commented:
What's the behavior?

Don't think the link is necessary.

Apercu commented on examples/complex-chart/package.json:
```diff
@@ -0,0 +1,28 @@
+{
```
The packages and lockfiles are gonna be the same too

Apercu commented on examples/webpack.config.js:
```diff
@@ -0,0 +1,37 @@
+  module: {
+    rules: [
+      {
+        test: /\.scss/,
```
Re-add `$` at the end

Apercu commented on yarn.lock:
```diff
@@ -1,5 +1,7 @@
 # THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Shouldn't have been touched

amilajack replied:
This changed when i installed the root dependencies. I will revert the changes made to this file.

Apercu commented on examples/webpack.config.js:
```diff
@@ -0,0 +1,37 @@
+// NOTE: This is a Webpack 2 configuration file for react-vis
```
I think we know that ;)

Apercu commented on examples/webpack.config.js:
```diff
@@ -0,0 +1,37 @@
+// NOTE: This is a Webpack 2 configuration file for react-vis
+const {resolve} = require('path');
```
Shouldn't need the resolve, a string will be enough.

amilajack commented:
> What's the behavior?

@Apercu If the user does not install the root dependencies, the following error will be thrown:

```
ERROR in ./app.js
Module build failed: ReferenceError: Unknown plugin "module-resolver" specified in ...
```

Apercu commented:
You should to specify a babel config in the `package.json`, otherwise it will try to find it in parent directories.

amilajack commented:
@Apercu Ahh i see. That works. I was under the impression that babel-loader required the babel config to be inside of the webpack config itself. I tested adding the config to package.json and it works! I'll remove the babel config from the webpack config.

mcnuttandrew commented on examples/examples.scss:
```diff
@@ -0,0 +1,287 @@
+@import "../../src/main.scss";
```
please scss-lint this file

amilajack replied:
Is scss-lint a tool that react-vis uses? I can't seem to find the dependency or npm script for adding sass linting, so I assume that it requires some kind of global installation to run. I can PR for [stylelint](https://github.com/stylelint/stylelint) later. Would that be welcome?

mcnuttandrew replied:
That's a wonderful idea! I would be really excited to see that PR. Style linting can wait for that pr in that case

amilajack commented:
Is this PR ready to be merged?

Apercu commented on examples/complex-chart/app.js:
```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
```
I'm not sure if we still need the header on every file, as @ibgreen removed it from the contribution guidelines, need confirmation

amilajack replied:
@mcnuttandrew does this need to be in every header file?

mcnuttandrew replied:
I'm gonna error on the side of caution and say that yes it must be in every file

Apercu commented on examples/complex-chart/complex-chart-example.js:
```diff
@@ -0,0 +1,199 @@
+
+export default class Example extends React.Component {
+
+  constructor(props) {
```
Remove the `constructor` and the `componentWillMount` and just use the state class property. [eg](https://github.com/uber/react-vis/blob/master/showcase/radial-chart/custom-radius-radial-chart.js#L27)

amilajack commented:
Made all the requested changes. Since @mcnuttandrew said that header should be in every file, is this ready to merge?

Apercu commented on examples/complex-chart/complex-chart-example.js:
```diff
@@ -0,0 +1,195 @@
+    }
+  ]
+
+  state = {
```
```
state = {
  crosshairValues: [],
  series: [
    {
      title: 'Apples',
      disabled: false,
      data: getRandomSeriesData(Math.random() * 50)
    },
    {
      title: 'Bananas',
      disabled: false,
      data: getRandomSeriesData(Math.random() * 50)
    }
  ]
}
```

Apercu commented:
Just need to change what I commented, maybe just put the `Math.random() * 50` in a constant at the top of the file

amilajack commented:
Somehow I missed the comment on that. Finished the changes

Apercu commented:
I don't see the change
